redis/src/commands/cluster_scan.rs

```diff
@@ -0,0 +1,709 @@
+    ) -> Self {
+        Self {
+            cursor,
+            scanned_slots_map,
+            address_in_scan,
+            address_epoch,
+        }
+    }
+
+    /// Initiate a scan operation by creating a new ScanState.
+    ///
+    /// This method creates a new ScanState with the initial values for a scan operation.
+    ///
+    /// # Arguments
+    ///
+    /// * `connection` - The connection to the Redis cluster.
+    ///
+    /// # Returns
+    ///
+    /// The initial ScanState.
+    pub(crate) async fn initiate_scan<C: ClusterInScan + ?Sized>(
+        connection: &C,
+    ) -> RedisResult<ScanState> {
+        let new_scanned_slots_map: SlotsBitsArray = [0; BITS_ARRAY_SIZE];
+        let new_cursor = 0;
+        let address_in_scan = connection.get_address_by_slot(0).await;
+        match address_in_scan {
+            Ok(address) => {
+                let address_epoch = connection.get_address_epoch(&address).await.unwrap_or(0);
+                Ok(ScanState::create(
+                    new_cursor,
+                    new_scanned_slots_map,
+                    address,
+                    address_epoch,
+                ))
+            }
+            Err(err) => Err(err),
+        }
+    }
+
+    /// Get the next slot that hasn't been scanned.
+    ///
+    /// This method iterates over the scanned slots map and finds the next slot that is set to 0.
+    ///
+    /// # Arguments
+    ///
+    /// * `scanned_slots_map` - The scanned slots map.
+    ///
+    /// # Returns
+    ///
+    /// The next slot that hasn't been scanned, or `None` if all slots have been scanned.
+    pub fn get_next_slot(&self, scanned_slots_map: &SlotsBitsArray) -> Option<u16> {
+        let all_slots_scanned = scanned_slots_map.iter().all(|&word| word == u64::MAX);
+        if all_slots_scanned {
+            return Some(END_OF_SCAN);
+        }
+        for (i, slot) in scanned_slots_map.iter().enumerate() {
+            let mut mask = 1;
+            for j in 0..BITS_PER_U64 {
+                if (slot & mask) == 0 {
+                    return Some((i * BITS_PER_U64 + j) as u16);
+                }
+                mask <<= 1;
+            }
+        }
+        None
+    }
+
+    /// Update the next address to be scanned without updating the scanned slots map.
+    /// This method is used when the address epoch has changed and we can't know which slots are new or when slots are missing.
+    /// In this case we will skip updating the scanned_slots_map and will just update the address and the cursor.
+    ///
+    /// # Arguments
+    ///
+    /// * `connection` - The connection to the Redis cluster.
+    ///
+    /// # Returns
+    ///
+    /// The updated ScanState.
+    pub(crate) async fn update_scan_state_without_updating_scanned_map<
+        C: ClusterInScan + ?Sized,
+    >(
+        &self,
+        connection: &C,
+    ) -> RedisResult<ScanState> {
+        let next_slot = self.get_next_slot(&self.scanned_slots_map).unwrap();
```

Comment:
`unwrap` in production code?

change this:

```rust
let next_slot = self.get_next_slot(&self.scanned_slots_map)?;
```

Reply:
Its an option and not result. and it cant be wrong, it wrapped in option for those principals. Which seems to me a bit redundant, you add option since its the principal, then you need to handle it in another place etc.
But ill add unwrap_or to avoid breaking the principals :).